docs/notebooks/jax-101/01-jax-basics.md

```diff
@@ -0,0 +1,375 @@
+
++++ {"id": "LNjf9jUEsZZ8"}
+
+## Value and Grad
+
+Often, you need to find both the value and the gradient of a function, e.g. if you want to log the training loss. JAX has a handy sister transformation for efficiently doing that:
+
+```{code-cell} ipython3
+:id: dWg4_-h3sYwl
+
+jax.value_and_grad(sum_squared_error)(x, y)
+```
+
++++ {"id": "QVT2EWHJsvvv"}
+
+which returns a tuple of, you guessed it, (value, grad). To be precise, for any `f`,
+
+```
+jax.value_and_grad(f)(*xs) == (f(*xs), jax.grad(f)(*xs)) 
+```
+
++++ {"id": "QmHTVpAks3OX"}
+
+## Auxiliary data
+
+In addition to wanting to log the value, we often want to report some intermediate results obtained in computing the loss function. But if we try doing that with regular `jax.grad`, we run into trouble:
+
+```{code-cell} ipython3
+:id: ffGCEzT4st41
+:tags: [raises-exception]
+
+def squared_error_with_aux(x, y):
+  return sum_squared_error(x, y), x-y
+
+jax.grad(squared_error_with_aux)(x, y)
+```
+
++++ {"id": "IUubno3nth4i"}
+
+This is because `jax.grad` is only defined on scalar functions, and our new function returns a tuple. But we need to return a tuple to return our intermediate results! This is where `has_aux` comes in:
+
+```{code-cell} ipython3
+:id: uzUFihyatgiF
+
+jax.grad(squared_error_with_aux, has_aux=True)(x, y)
+```
+
++++ {"id": "g5s3UiFauwDk"}
+
+`has_aux` signifies that the function returns a pair, `(out, aux)`. It makes `jax.grad` ignore `aux`, passing it through to the user, while differentiating the function as if only `out` was returned.
+
++++ {"id": "fk4FUXe7vsW4"}
+
+## Differences from Numpy
+
+The `jax.numpy` API closely follows that of Numpy. However, there are some important differences. We cover many of these in future guides, but it's worth pointing some out now.
+
+The most important difference, and in some sense the root of all the rest, is that JAX is designed to be _functional_, as in _functional programming_. The reason behind this is that the kinds of program transformations that JAX enables are much more feasible in functional-style programs.
+
+An introduction to functional programming (FP) is out of scope of this guide. If you already are familiar with FP, you will find your FP intuition helpful while learning JAX. If not, don't worry! The important feature of functional programming to grok when working with JAX is very simple: don't write code with side-effects.
+
+A side-effect is any effect of a function that doesn't appear in its output. One example is modifying an array in place:
+
+```{code-cell} ipython3
+:id: o_YBuLQC1wPJ
+
+import numpy as np
+
+x = np.array([1, 2, 3])
+
+def in_place_modify(x):
+  x[0] = 123
+  return None
+
+in_place_modify(x)
+x
+```
+
++++ {"id": "JTtUihVZ13F6"}
+
+The side-effectful function modifies its argument, but returns a completely unrelated value. The modification is a side-effect. 
+
+This will run in Numpy. However, JAX arrays won't allow themselves to be modified in-place:
```

Comment:
Suggestion:

```suggestion
This will run in NumPy. However, JAX arrays won't allow themselves to be modified in-place:
```

Reply:
Suggestion:

Also, if the "this" refers to the code below, maybe it's best to say "The code below will run on NumPy"

```suggestion
The code below will run in NumPy. However, JAX arrays won't allow themselves to be modified in-place:
```